### PR TITLE
Add FilterToolbar to migration plan table

### PR DESCRIPTION
**Pull request by gildub**
Resolves https://github.com/konveyor/virt-ui/issues/76

### CODE REVIEW

**Comment by gildub** [comment]
@vconzola, besides plan migration `name` are there any other fields needing to have filter option?

**Comment by vconzola** [comment]
@gildub Yes, user should be able to filter by Source provider, Target provider, and Plan status.

File: src/app/Plans/components/PlansTable.tsx
Comment (mturley) [inline src/app/Plans/components/PlansTable.tsx:47]: Per https://github.com/konveyor/virt-ui/issues/76#issuecomment-697908751 we'll need to add filter categories for all columns except number of VMs. so I think it would be Search filters for source provider / target provider (although maybe Select instead since we can query the list of available providers), and a Select filter for the plan status (logic might be tricky there... I can't remember if I made the Select filter able to have custom filter logic and not just a plain string match. if that's hard we could wait for another PR).

@vconzola should the filters for source/target provider in the plans table be search filters or select/dropdown filters?

File: src/app/Plans/components/PlansTable.tsx
Comment (mturley) [inline src/app/Plans/components/PlansTable.tsx:195]: The placement of the Create button in the dead center of the page is a little weird, I think we should put it in that first LevelItem right next to the FilterToolbar. We probably want it vertically-aligned at the top instead of the middle (so it stays in the same place visually when the filter chips appear) and with a medium margin from the toolbar, but I'm not sure what will be needed exactly until we move it.

File: src/app/common/hooks/useFilterState.ts
Comment (mturley) [inline src/app/common/hooks/useFilterState.ts:25]: Why this change? it was case-insensitive before due to the toLowerCase calls.

**Review by mturley — CHANGES_REQUESTED** [review]

File: src/app/common/hooks/useFilterState.ts
Comment (mturley) [reply src/app/common/hooks/useFilterState.ts:25]: Ah I see, is this to support the Select filter logic? If so we might want to have two different branches here depending on the filter type, so the Search ones can still be case-insensitive

File: src/app/Plans/components/PlansTable.tsx
Comment (vconzola) [reply src/app/Plans/components/PlansTable.tsx:195]: @mturley @gildub The Create button should be 16 px to the right of the search field as documented here: https://www.patternfly.org/v4/components/toolbar/react

File: src/app/Plans/components/PlansTable.tsx
Comment (vconzola) [reply src/app/Plans/components/PlansTable.tsx:47]: @mturley Not sure what you mean by search filters or select/dropdown filters, but they should be this: https://www.patternfly.org/v4/guidelines/filters#attribute-value-textbox-filters

Comment (mturley) [reply src/app/Plans/components/PlansTable.tsx:47]: @vconzola what we're calling a "search filter" is the first mockup shown at that link, where the value is typed in a text box. What we're calling a "select filter" is the second mockup there, where the value is chosen from a list.

I wasn't sure which would be better since provider names are plain text, but the list of possible values is known.

Comment (vconzola) [reply src/app/Plans/components/PlansTable.tsx:47]: @mturley Use what you're calling a search filter.

Comment (mturley) [reply src/app/Plans/components/PlansTable.tsx:47]: Got it, thanks

Comment (gildub) [reply src/app/Plans/components/PlansTable.tsx:47]: @mturley, sorry I totally zapped the comment on #76.

Adding them.

File: src/app/common/hooks/useFilterState.ts
Comment (gildub) [reply src/app/common/hooks/useFilterState.ts:25]: Ok I didn't know the requirement the filter has to be case insensitive. I'll restore it.

Comment (mturley) [reply src/app/common/hooks/useFilterState.ts:25]: I suppose it's not a hard requirement, I just thought it'd be nice. If it's difficult don't worry about it.

Comment (gildub) [reply src/app/common/hooks/useFilterState.ts:25]: Not hard, just oversimplification shortcut I shouldn't taken.

File: src/app/Plans/components/PlansTable.tsx
Comment (gildub) [reply src/app/Plans/components/PlansTable.tsx]: @vconzola, I've re-positioned button. Please validate.

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (mturley) [inline src/app/common/components/FilterToolbar/FilterToolbar.tsx:35]: This should be `ISelectFilterCategory<T> extends IBasicFilterCategory<T>`, the `IPlan` can be passed in when it's used.

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (mturley) [inline src/app/common/components/FilterToolbar/FilterToolbar.tsx:39]: Same here

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (gildub) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx:35]: I needs a real type here because it's passing as parameter to the `IBasicFilterCategory` interface

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (gildub) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx:39]: same

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (mturley) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx:35]: Right, ok. The FilterToolbar file shouldn't be specific to Plans though, and `ISelectFilterCategory<T> extends IBasicFilterCategory<T>` will define this interface as a generic one. You can then use `ISelectFilterCategory<IPlan>` in the actual table component and that will define what `T` means here, so it will make its way to the `IBasicFilterCategory`.

Comment (mturley) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx:35]: To get this to work you will actually also need to add generics to `FilterCategory`, turn it into `export type FilterCategory<T> = ISearchFilterCategory<T> | ISelectFilterCategory<T>;` and then use `FilterCategory<IPlan>[]` as the type of the categories array in the table. That should get you the right types passed down.

Comment (gildub) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx]: Works fine!

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (gildub) [reply src/app/common/components/FilterToolbar/FilterToolbar.tsx]: same same, thanks

File: src/app/common/components/FilterToolbar/FilterControl.tsx
Comment (mturley) [inline src/app/common/components/FilterToolbar/FilterControl.tsx:12]: We also shouldn't have `IPlan` in here, since FilterControl is also supposed to be generic. I know you had to leave for the day, I'll just push a commit to fix it and merge.

File: src/app/common/components/FilterToolbar/FilterToolbar.tsx
Comment (mturley) [inline src/app/common/components/FilterToolbar/FilterToolbar.tsx:50]: Same here, we shouldn't have `IPlan` anywhere in `common/components/*`

File: src/app/common/hooks/useFilterState.ts
Comment (mturley) [inline src/app/common/hooks/useFilterState.ts:11]: This was fine with `<T>`, we just have to thread the `T` down through all the interfaces with generics at every level. `IPlan` should only be in the table.

**Review by mturley — COMMENTED** [review]
Everything looks good except for the final feedback about generics, but I don't want to hold this up. I'll push a commit to fix it in a moment, and we can look at it together after we merge.

**Comment by mturley** [comment]
Ok @gildub the trick was to make everything that deals with the filter categories generic, including the components themselves. I figured out how to do that via https://wanago.io/2020/03/09/functional-react-components-with-generic-props-in-typescript/, basically the pattern is weird because we can't use `React.FunctionComponent` with a generic type interface. So we use some types it uses internally (`PropsWithChildren` and `JSX.Element`... I know I said basically never use it but this is an exception because it's what React expects a function component to return)

This is the basic pattern for a generic component that I arrived at, it's just like any other generic function:
```tsx
// Definition:
const MyComponent = <T>(props: React.PropsWithChildren<IMyComponentProps<T>>): JSX.Element | null => ...
// Usage:
<MyComponent<ISomething> some="props">
```

I think this is a really cool pattern we can use in a bunch of places for lib-ui! It's a shame the type annotation is so awkward though.

**Review by mturley — APPROVED** [review]
LGTM!

**Comment by gildub** [comment]
@mturley, that's really nice. Well done.

File: src/app/common/components/FilterToolbar/FilterControl.tsx
Comment (gildub) [reply src/app/common/components/FilterToolbar/FilterControl.tsx]: Yeah exactly, that was not ideal but working!

**Pull request merged**
